### ISSUE TITLE
An error occurs when training BinaryNet on mnist dataset

### ISSUE THREAD

**Issue by bohanzhuang** [MatthieuCourbariaux/BinaryNet#2]
I'm trying to train Binarynet on the mnist dataset, but it prints the results: 

Training...
Epoch 1 of 1000 took 70.4168188572s
  LR:                            0.003
  training loss:                 nan
  validation loss:               nan
  validation error rate:         90.0900000334%
  best epoch:                    1
  best validation error rate:    90.0900000334%
  test loss:                     nan
  test error rate:               90.2000000477%
Epoch 2 of 1000 took 69.7526259422s
  LR:                            0.00297249583468
  training loss:                 nan
  validation loss:               nan
  validation error rate:         90.0900000334%
  best epoch:                    2
  best validation error rate:    90.0900000334%
  test loss:                     nan
  test error rate:               90.2000000477%

I don't know what's going wrong....So anyone has some ideas of solving it?

**Comment by MatthieuCourbariaux** [comment]
Here is what I get on my machine:

Training...
Epoch 1 of 1000 took 27.6782290936s
  LR:                            0.003
  training loss:                 0.237839879617
  validation loss:               0.0515174402669
  validation error rate:         6.72999999858%
  best epoch:                    1
  best validation error rate:    6.72999999858%
  test loss:                     0.052927171737
  test error rate:               7.02999999747%
...

Did you bring any modification to the code?
Are you using the latest version of Theano, Lasagne and Pylearn2?
Did you get your MNIST dataset from http://yann.lecun.com/exdb/mnist/ ?
Do you get any warning?
What GPU are you using?
What are your Theano flags?

**Comment by bohanzhuang** [comment]
@MatthieuCourbariaux  Thanks for your kindly reply!

1: No, I didn't modify the code. 
2: I installed the Bleeding-edge version lasagne according to the official guide:
 pip install --upgrade https://github.com/Theano/Theano/archive/master.zip
 pip install --upgrade https://github.com/Lasagne/Lasagne/archive/master.zip
Theano and lasagne are the lastest. 
3: Yes, I directly downloaded the MNIST dataset. 
4: The only warning is 
/usr/local/lib/python2.7/dist-packages/theano/tensor/signal/downsample.py:5: UserWarning: downsample module has been moved to the pool module.
  warnings.warn("downsample module has been moved to the pool module.")
/usr/local/lib/python2.7/dist-packages/theano/sandbox/cuda/nvcc_compiler.py:224: UserWarning: You have the cuda library directory in your lib_dirs. This has been known to cause problems and should not be done.
  warnings.warn("You have the cuda library directory in your "

I think this doesn't matter. 
5: GeForce GTX TITAN X
6: Theano flags: 
[global]
floatX = float32
optimizer = fast_compile
[nvcc]
fastmath = True
[cuda]
root = /usr/local/cuda-7.5

**Comment by MatthieuCourbariaux** [comment]
OK, thank you for the details,

You should try without "optimizer = fast_compile" in your .theanorc
I think it might be the cause of your problem.

**Comment by bohanzhuang** [comment]
@MatthieuCourbariaux   Perfect, solved!
